### PR TITLE
wip adding coverage

### PR DESCRIPTION
**Pull request by nh13**
@lihaoyi I am trying to add a coverage module like [`sbt-coverage`](https://github.com/scoverage/sbt-scoverage) right into `mill`.  If you think it's better to have a separate repo, rather than putting it into `mill` let me know.  I was thinking you already had support for tests in there.

There are a number of remaining issues that are likely due to my beginner understanding of `mill` that I could use help with.

1. I have had to hard-code the `scoverage` version into the mill build file.  I don't see a way around this.  You can modify the version of the plugin run on the command line, but who's to say that it will always be compatible with the one used in the mill build file.  Thoughts?
2. I'd like to add a `trait` into `ScalaModule` similar to `Tests`, but I get cyclic dependency errors with this: `trait Coverage extends CoverageModule with Tests`.
3. To set the plugin data path correctly (where the coverage information gets written), I need to pass the path while while setting the `scalacOptions`.   I have hard-coded those paths (see `coverageDataDir` and `coverageRootDir`), which I'd like to change but don't know how.  I also want to get the output location (`*dest`) of the `coverage` target but don't know how.
4. The `AggregateCoverageModule` is meant to aggregate coverage data across different modules.  I have a `def modules` so the user can set the modules to aggregate, but I don't see a way to get the list of sources across all those modules (see `AggregateCoverageModule.sources`, which doesn't work).  I'd also like to get the source encoding for each (see the `FIXME: sourceEncoding(scalacOptions())`).

I know I am asking for a bunch of help here, but I think coverage reports will be very useful for folks.  Once I get some these technical challenges out of they way, I can add documentation and write some tests.  Thanks for any insights and help in advance.

### CODE REVIEW

File: build.sc [hunk 1]
```diff
@@ -137,7 +137,8 @@ object scalalib extends MillModule {
   def moduleDeps = Seq(main)
 
   def ivyDeps = Agg(
-    ivy"org.scala-sbt:test-interface:1.0"
+    ivy"org.scala-sbt:test-interface:1.0",
+    ivy"org.scoverage::scalac-scoverage-plugin:1.4.0-M3"
```
Comment (nh13) [inline build.sc]: scoverage plugin version is hard-coded

File: scalalib/src/mill/scalalib/CoverageModule.scala [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+      Option(excludedFiles).map(v => s"-P:scoverage:excludedFiles:$v"),
+      if (highlighting) Some("-Yrangepos") else None // rangepos is broken in some releases of scala so option to turn it off
+    ).flatten
+  }
+
+  private def coverageRootDir = T {
+    // FIXME: using artifactName is wrong
+    val path = millSourcePath / 'out / artifactName()
+      .split('-')
+      .last / scalaVersion() / 'coverage / 'dest
```
Comment (nh13) [inline scalalib/src/mill/scalalib/CoverageModule.scala]: Finding the destination path for the `coverage` target (or whatever the user will call it) would be nice.

File: scalalib/src/mill/scalalib/CoverageModule.scala [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+  }
+}
+
+trait AggregateCoverageModule extends Module with CoverageOptions {
+  import CoverageUtil._
+
+  def modules: Seq[TestModule]
+
+  // FIXME: I want the sources!
+  def sources = T { modules.flatMap { module => module.sources() } }
```
Comment (nh13) [inline scalalib/src/mill/scalalib/CoverageModule.scala]: How do I get the aggregate sources across modules?

File: scalalib/src/mill/scalalib/CoverageModule.scala [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+  def aggregate = T {
+    val log = T.ctx.log
+    log.info(s"Aggregating coverage from subprojects...")
+
+    val rootDir = (millSourcePath / 'dest).toNIO
+    val encoding: Option[String] = None // FIXME: sourceEncoding(scalacOptions())
```
Comment (nh13) [inline scalalib/src/mill/scalalib/CoverageModule.scala]: Ditto, how do I get the source encoding across modules (I can make sure they are all the same if necessary).

File: scalalib/src/mill/scalalib/CoverageModule.scala [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+/** A little trait to mix in within a module to add coverage.
+  *
+  * {{{
+  * object coverage extends Tests with CoverageModule {
+  *     def testIvyDeps = Agg(ivy"org.scalatest::scalatest:3.0.1")
+  *     def testFrameworks = Seq("org.scalatest.tools.Framework")
+  * }
+  * }}}
```
Comment (nh13) [inline scalalib/src/mill/scalalib/CoverageModule.scala]: I'd like to figure out how to add `trait Coverage extends CoverageModule with Tests` to `ScalaModule`.

File: scalalib/src/mill/scalalib/CoverageModule.scala [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        .getOrElse {
+          throw new Exception(
+            s"Fatal: ${ScalacPluginArtifact} (${scoverageVersion}) not found on the classpath:\n\t" + cp
+              .map(_.path)
+              .mkString("\n\t"))
+        }
+    }.path
+    super.scalacOptions() ++ Seq(
+      Some(s"-Xplugin:${pluginPath}"),
+      Some(s"-P:scoverage:dataDir:${coverageDataDir().toNIO}"),
```
Comment (nh13) [inline scalalib/src/mill/scalalib/CoverageModule.scala]: Here's where the path to the coverage output directory is set.

**Comment by nh13** [comment]
@lihaoyi any thoughts on this PR? I have a need for coverage information, and I'd rather it be _native_ in mill.  Thoughts?

File: scalalib/src/mill/scalalib/CoverageModule.scala (hunk 3 above)
Comment (lihaoyi) [reply scalalib/src/mill/scalalib/CoverageModule.scala]: See how [transitiveIvyDeps](https://github.com/lihaoyi/mill/blob/master/scalalib/src/mill/scalalib/JavaModule.scala#L84) doe this

File: scalalib/src/mill/scalalib/CoverageModule.scala (hunk 4 above)
Comment (lihaoyi) [reply scalalib/src/mill/scalalib/CoverageModule.scala]: See [transitiveIvyDeps](https://github.com/lihaoyi/mill/blob/master/scalalib/src/mill/scalalib/JavaModule.scala#L84)

File: scalalib/src/mill/scalalib/CoverageModule.scala (hunk 5 above)
Comment (lihaoyi) [reply scalalib/src/mill/scalalib/CoverageModule.scala]: No idea what this means

**Comment by lihaoyi** [comment]
no idea, you'll have to figure it out

**Comment by lihaoyi** [comment]
@nh13 what is the status on this?

**Comment by nh13** [comment]
I am debugging the following:

1. I run `ci/publish-local.sh` to create a standalone assembly. 
2. I run `~/mill-release scalacoverage.test` and get the following:

<details>

```
[279/279] scalacoverage.test.test 
-------------------------------- Running Tests --------------------------------
[44/69] buildUTest[2.11.12,1.4.0-M3].compile 
[info] Compiling 1 Scala source to /Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildUTest/2.11.12/1.4.0-M3/compile/dest/classes ...
[info] Ignoring duplicate plugin scoverage (scoverage.ScoveragePlugin)
[info] [info] Cleaning datadir [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildUTest/2.11.12/1.4.0-M3/coverageDataDir/dest]
[info] [info] Beginning coverage instrumentation
[info] [info] Instrumentation completed [14 statements]
[info] [info] Wrote instrumentation file [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildUTest/2.11.12/1.4.0-M3/coverageDataDir/dest/scoverage.coverage.xml]
[info] [info] Will write measurement data to [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildUTest/2.11.12/1.4.0-M3/coverageDataDir/dest]
[info] Done compiling.
[63/69] buildUTest[2.11.12,1.4.0-M3].test.compile 
[info] Compiling 1 Scala source to /Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildUTest/2.11.12/1.4.0-M3/test/compile/dest/classes ...
[info] Ignoring duplicate plugin scoverage (scoverage.ScoveragePlugin)
[info] [info] Cleaning datadir [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildUTest/2.11.12/1.4.0-M3/coverageDataDir/dest]
[info] [info] Beginning coverage instrumentation
[info] [info] Instrumentation completed [0 statements]
[info] [info] Wrote instrumentation file [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildUTest/2.11.12/1.4.0-M3/coverageDataDir/dest/scoverage.coverage.xml]
[info] [info] Will write measurement data to [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildUTest/2.11.12/1.4.0-M3/coverageDataDir/dest]
[info] Done compiling.
[68/69] buildUTest[2.11.12,1.4.0-M3].test.forkWorkingDir 
mill.eval.Evaluator.$anonfun$evaluateGroup$15(Evaluator.scala:317)
scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
scala.Console$.withErr(Console.scala:192)
mill.eval.Evaluator.$anonfun$evaluateGroup$14(Evaluator.scala:315)
scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
scala.Console$.withOut(Console.scala:163)
mill.eval.Evaluator.$anonfun$evaluateGroup$13(Evaluator.scala:314)
scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
scala.Console$.withIn(Console.scala:226)
mill.eval.Evaluator.$anonfun$evaluateGroup$9(Evaluator.scala:313)
mill.eval.Evaluator.$anonfun$evaluateGroup$9$adapted(Evaluator.scala:269)
scala.collection.Iterator.foreach(Iterator.scala:944)
scala.collection.Iterator.foreach$(Iterator.scala:944)
scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
scala.collection.IterableLike.foreach(IterableLike.scala:71)
scala.collection.IterableLike.foreach$(IterableLike.scala:70)
scala.collection.AbstractIterable.foreach(Iterable.scala:54)
mill.eval.Evaluator.evaluateGroup(Evaluator.scala:269)
mill.eval.Evaluator.evaluateGroupCached(Evaluator.scala:175)
mill.eval.Evaluator.$anonfun$evaluate$2(Evaluator.scala:57)
mill.eval.Evaluator.$anonfun$evaluate$2$adapted(Evaluator.scala:49)
scala.collection.Iterator.foreach(Iterator.scala:944)
scala.collection.Iterator.foreach$(Iterator.scala:944)
scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
mill.eval.Evaluator.evaluate(Evaluator.scala:49)
mill.util.TestEvaluator.apply(TestEvaluator.scala:35)
mill.scalacoverage.ScoverageReportModuleTest$.runTests$1(ScoverageReportModuleTest.scala:69)
mill.scalacoverage.ScoverageReportModuleTest$.checkUtest$1(ScoverageReportModuleTest.scala:88)
mill.scalacoverage.ScoverageReportModuleTest$.$anonfun$tests$15(ScoverageReportModuleTest.scala:108)
utest.framework.StackMarker$.dropOutside(StackMarker.scala:13)
utest.framework.TestCallTree.run(Model.scala:35)
utest.framework.TestCallTree.run(Model.scala:33)
utest.framework.TestCallTree.run(Model.scala:33)
utest.framework.TestCallTree.run(Model.scala:33)
utest.TestRunner$.$anonfun$runAsync$5(TestRunner.scala:74)
utest.framework.Executor.utestWrap(Executor.scala:12)
utest.framework.Executor.utestWrap$(Executor.scala:10)
utest.TestSuite.utestWrap(TestSuite.scala:18)
utest.TestRunner$.$anonfun$runAsync$4(TestRunner.scala:71)
utest.framework.StackMarker$.dropOutside(StackMarker.scala:13)
utest.TestRunner$.$anonfun$runAsync$2(TestRunner.scala:71)
utest.TestRunner$.evaluateFutureTree(TestRunner.scala:171)
utest.TestRunner$.$anonfun$evaluateFutureTree$2(TestRunner.scala:174)
scala.concurrent.Future$.$anonfun$traverse$1(Future.scala:845)
scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:56)
scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:64)
scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:48)
scala.concurrent.Future$.traverse(Future.scala:845)
utest.TestRunner$.evaluateFutureTree(TestRunner.scala:174)
utest.TestRunner$.$anonfun$evaluateFutureTree$2(TestRunner.scala:174)
scala.concurrent.Future$.$anonfun$traverse$1(Future.scala:845)
scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:56)
scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:64)
scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:48)
scala.concurrent.Future$.traverse(Future.scala:845)
utest.TestRunner$.evaluateFutureTree(TestRunner.scala:174)
utest.TestRunner$.$anonfun$evaluateFutureTree$2(TestRunner.scala:174)
scala.concurrent.Future$.$anonfun$traverse$1(Future.scala:845)
scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:56)
scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:64)
scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:48)
scala.concurrent.Future$.traverse(Future.scala:845)
utest.TestRunner$.evaluateFutureTree(TestRunner.scala:174)
utest.TestRunner$.runAsync(TestRunner.scala:99)
utest.runner.BaseRunner.runSuite(BaseRunner.scala:170)
utest.runner.BaseRunner.$anonfun$makeTask$1(BaseRunner.scala:181)
utest.runner.Task.execute(Task.scala:20)
mill.scalalib.TestRunner$.$anonfun$runTests$2(TestRunner.scala:93)
scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
scala.collection.TraversableLike.map(TraversableLike.scala:234)
scala.collection.TraversableLike.map$(TraversableLike.scala:227)
scala.collection.AbstractTraversable.map(Traversable.scala:104)
mill.scalalib.TestRunner$.$anonfun$runTests$1(TestRunner.scala:76)
mill.modules.Jvm$.inprocess(Jvm.scala:135)
mill.scalalib.TestRunner$.runTests(TestRunner.scala:71)
mill.scalalib.TestRunner$.main(TestRunner.scala:49)
mill.scalalib.TestRunner.main(TestRunner.scala)
X mill.scalacoverage.ScoverageReportModuleTest.scalacoverage.coverageReport.utest_scala_2_11_12_scoverage_1_4 6570ms 
  java.lang.NullPointerException
    coursier.core.Parse$.versionInterval(Parse.scala:60)
    coursier.ivy.IvyRepository.find(IvyRepository.scala:151)
    coursier.Fetch$.$anonfun$find$1(Fetch.scala:38)
    scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
    scala.collection.immutable.List.foreach(List.scala:389)
    scala.collection.TraversableLike.map(TraversableLike.scala:234)
    scala.collection.TraversableLike.map$(TraversableLike.scala:227)
    scala.collection.immutable.List.map(List.scala:295)
    coursier.Fetch$.find(Fetch.scala:38)
    coursier.Fetch$.get$1(Fetch.scala:67)
    coursier.Fetch$.$anonfun$from$2(Fetch.scala:68)
    scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
    scala.collection.Iterator.foreach(Iterator.scala:944)
    scala.collection.Iterator.foreach$(Iterator.scala:944)
    scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
    scala.collection.IterableLike.foreach(IterableLike.scala:71)
    scala.collection.IterableLike.foreach$(IterableLike.scala:70)
    scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    scala.collection.TraversableLike.map(TraversableLike.scala:234)
    scala.collection.TraversableLike.map$(TraversableLike.scala:227)
    scala.collection.AbstractTraversable.map(Traversable.scala:104)
    coursier.Fetch$.$anonfun$from$1(Fetch.scala:65)
    coursier.core.ResolutionProcess$.$anonfun$fetchAll$10(ResolutionProcess.scala:197)
    coursier.util.Task$.$anonfun$flatMap$2(Task.scala:10)
    scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:303)
    scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
    scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
    java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
    java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
    java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
    java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[44/69] buildScalaTest[2.11.12,1.4.0-M3].compile 
[info] Compiling 1 Scala source to /Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildScalaTest/2.11.12/1.4.0-M3/compile/dest/classes ...
[info] Ignoring duplicate plugin scoverage (scoverage.ScoveragePlugin)
[info] [info] Cleaning datadir [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildScalaTest/2.11.12/1.4.0-M3/coverageDataDir/dest]
[info] [info] Beginning coverage instrumentation
[info] [info] Instrumentation completed [14 statements]
[info] [info] Wrote instrumentation file [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildScalaTest/2.11.12/1.4.0-M3/coverageDataDir/dest/scoverage.coverage.xml]
[info] [info] Will write measurement data to [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildScalaTest/2.11.12/1.4.0-M3/coverageDataDir/dest]
[info] Done compiling.
[63/69] buildScalaTest[2.11.12,1.4.0-M3].test.compile 
[info] Compiling 1 Scala source to /Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildScalaTest/2.11.12/1.4.0-M3/test/compile/dest/classes ...
[info] Ignoring duplicate plugin scoverage (scoverage.ScoveragePlugin)
[info] [info] Cleaning datadir [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildScalaTest/2.11.12/1.4.0-M3/coverageDataDir/dest]
[info] [info] Beginning coverage instrumentation
[info] [info] Instrumentation completed [12 statements]
[info] [info] Wrote instrumentation file [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildScalaTest/2.11.12/1.4.0-M3/coverageDataDir/dest/scoverage.coverage.xml]
[info] [info] Will write measurement data to [/Users/nhomer/git/nh13/mill/target/workspace/mill/scalacoverage/ScoverageReportModuleTest/basicCoverageEvaluator/buildScalaTest/2.11.12/1.4.0-M3/coverageDataDir/dest]
[info] Done compiling.
[68/69] buildScalaTest[2.11.12,1.4.0-M3].test.forkWorkingDir 
mill.eval.Evaluator.$anonfun$evaluateGroup$15(Evaluator.scala:317)
scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
scala.Console$.withErr(Console.scala:192)
mill.eval.Evaluator.$anonfun$evaluateGroup$14(Evaluator.scala:315)
scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
scala.Console$.withOut(Console.scala:163)
mill.eval.Evaluator.$anonfun$evaluateGroup$13(Evaluator.scala:314)
scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
scala.Console$.withIn(Console.scala:226)
mill.eval.Evaluator.$anonfun$evaluateGroup$9(Evaluator.scala:313)
mill.eval.Evaluator.$anonfun$evaluateGroup$9$adapted(Evaluator.scala:269)
scala.collection.Iterator.foreach(Iterator.scala:944)
scala.collection.Iterator.foreach$(Iterator.scala:944)
scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
scala.collection.IterableLike.foreach(IterableLike.scala:71)
scala.collection.IterableLike.foreach$(IterableLike.scala:70)
scala.collection.AbstractIterable.foreach(Iterable.scala:54)
mill.eval.Evaluator.evaluateGroup(Evaluator.scala:269)
mill.eval.Evaluator.evaluateGroupCached(Evaluator.scala:175)
mill.eval.Evaluator.$anonfun$evaluate$2(Evaluator.scala:57)
mill.eval.Evaluator.$anonfun$evaluate$2$adapted(Evaluator.scala:49)
scala.collection.Iterator.foreach(Iterator.scala:944)
scala.collection.Iterator.foreach$(Iterator.scala:944)
scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
mill.eval.Evaluator.evaluate(Evaluator.scala:49)
mill.util.TestEvaluator.apply(TestEvaluator.scala:35)
mill.scalacoverage.ScoverageReportModuleTest$.runTests$1(ScoverageReportModuleTest.scala:69)
mill.scalacoverage.ScoverageReportModuleTest$.checkScalaTest$1(ScoverageReportModuleTest.scala:99)
mill.scalacoverage.ScoverageReportModuleTest$.$anonfun$tests$16(ScoverageReportModuleTest.scala:109)
utest.framework.StackMarker$.dropOutside(StackMarker.scala:13)
utest.framework.TestCallTree.run(Model.scala:35)
utest.framework.TestCallTree.run(Model.scala:33)
utest.framework.TestCallTree.run(Model.scala:33)
utest.framework.TestCallTree.run(Model.scala:33)
utest.TestRunner$.$anonfun$runAsync$5(TestRunner.scala:74)
utest.framework.Executor.utestWrap(Executor.scala:12)
utest.framework.Executor.utestWrap$(Executor.scala:10)
utest.TestSuite.utestWrap(TestSuite.scala:18)
utest.TestRunner$.$anonfun$runAsync$4(TestRunner.scala:71)
utest.framework.StackMarker$.dropOutside(StackMarker.scala:13)
utest.TestRunner$.$anonfun$runAsync$2(TestRunner.scala:71)
utest.TestRunner$.evaluateFutureTree(TestRunner.scala:171)
utest.TestRunner$.$anonfun$evaluateFutureTree$2(TestRunner.scala:174)
scala.concurrent.Future$.$anonfun$traverse$1(Future.scala:845)
scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:56)
scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:64)
scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:48)
scala.concurrent.Future$.traverse(Future.scala:845)
utest.TestRunner$.evaluateFutureTree(TestRunner.scala:174)
utest.TestRunner$.$anonfun$evaluateFutureTree$2(TestRunner.scala:174)
scala.concurrent.Future$.$anonfun$traverse$1(Future.scala:845)
scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:56)
scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:64)
scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:48)
scala.concurrent.Future$.traverse(Future.scala:845)
utest.TestRunner$.evaluateFutureTree(TestRunner.scala:174)
utest.TestRunner$.$anonfun$evaluateFutureTree$2(TestRunner.scala:174)
scala.concurrent.Future$.$anonfun$traverse$1(Future.scala:845)
scala.collection.IndexedSeqOptimized.foldLeft(IndexedSeqOptimized.scala:56)
scala.collection.IndexedSeqOptimized.foldLeft$(IndexedSeqOptimized.scala:64)
scala.collection.mutable.ArrayBuffer.foldLeft(ArrayBuffer.scala:48)
scala.concurrent.Future$.traverse(Future.scala:845)
utest.TestRunner$.evaluateFutureTree(TestRunner.scala:174)
utest.TestRunner$.runAsync(TestRunner.scala:99)
utest.runner.BaseRunner.runSuite(BaseRunner.scala:170)
utest.runner.BaseRunner.$anonfun$makeTask$1(BaseRunner.scala:181)
utest.runner.Task.execute(Task.scala:20)
mill.scalalib.TestRunner$.$anonfun$runTests$2(TestRunner.scala:93)
scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
scala.collection.TraversableLike.map(TraversableLike.scala:234)
scala.collection.TraversableLike.map$(TraversableLike.scala:227)
scala.collection.AbstractTraversable.map(Traversable.scala:104)
mill.scalalib.TestRunner$.$anonfun$runTests$1(TestRunner.scala:76)
mill.modules.Jvm$.inprocess(Jvm.scala:135)
mill.scalalib.TestRunner$.runTests(TestRunner.scala:71)
mill.scalalib.TestRunner$.main(TestRunner.scala:49)
mill.scalalib.TestRunner.main(TestRunner.scala)
X mill.scalacoverage.ScoverageReportModuleTest.scalacoverage.coverageReport.scalaTest_scala_2_11_12_scoverage_1_4 2324ms 
  java.lang.NullPointerException
    coursier.core.Parse$.versionInterval(Parse.scala:60)
    coursier.ivy.IvyRepository.find(IvyRepository.scala:151)
    coursier.Fetch$.$anonfun$find$1(Fetch.scala:38)
    scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
    scala.collection.immutable.List.foreach(List.scala:389)
    scala.collection.TraversableLike.map(TraversableLike.scala:234)
    scala.collection.TraversableLike.map$(TraversableLike.scala:227)
    scala.collection.immutable.List.map(List.scala:295)
    coursier.Fetch$.find(Fetch.scala:38)
    coursier.Fetch$.get$1(Fetch.scala:67)
    coursier.Fetch$.$anonfun$from$2(Fetch.scala:68)
    scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
    scala.collection.Iterator.foreach(Iterator.scala:944)
    scala.collection.Iterator.foreach$(Iterator.scala:944)
    scala.collection.AbstractIterator.foreach(Iterator.scala:1432)
    scala.collection.IterableLike.foreach(IterableLike.scala:71)
    scala.collection.IterableLike.foreach$(IterableLike.scala:70)
    scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    scala.collection.TraversableLike.map(TraversableLike.scala:234)
    scala.collection.TraversableLike.map$(TraversableLike.scala:227)
    scala.collection.AbstractTraversable.map(Traversable.scala:104)
    coursier.Fetch$.$anonfun$from$1(Fetch.scala:65)
    coursier.core.ResolutionProcess$.$anonfun$fetchAll$10(ResolutionProcess.scala:197)
    coursier.util.Task$.$anonfun$flatMap$2(Task.scala:10)
    scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:303)
    scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
    scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
    java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
    java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
    java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
    java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
    java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
1 targets failed
scalacoverage.test.test mill.scalacoverage.ScoverageReportModuleTest mill.scalacoverage.ScoverageReportModuleTest.scalacoverage.coverageReport.utest_scala_2_11_12_scoverage_1_4 and 1 more
```
</details>

I am guessing it's some dependency resolution issue.  I am following the same pattern as in `scalanativelib`, but I can't figure out why the latter works but not `scalacoverage`.  Unfortunately, the stacktrace is not helpful to find exactly what is null.

**Comment by nh13** [comment]
@lihaoyi I've spent another on this without getting any farther and having a clue on a path forward.

**Comment by DavidGregory084** [comment]
FWIW @nh13 I ran into this exact problem [here](https://github.com/lihaoyi/mill/blob/master/contrib/scalapblib/test/src/mill/contrib/scalapblib/TutorialTests.scala#L88) but it didn't occur when I used the `dev` projects to test it, only in the test evaluator I think

**Comment by nh13** [comment]
@DavidGregory084 "the dev projects to test it, only in the test evaluator I think".  What do you mean by the "dev projects " vs. the "test evaluator"?  I want to see if this problem goes away with the "dev projects" but I am not familiar with what that is.

**Comment by DavidGregory084** [comment]
Sorry I couldn't quite remember the detail of this when I wrote the original comment. What I was referring to was testing my changes on the project in the scratch folder of the mill repo, e.g.:

`mill -i dev.run scratch scalapb.compile`

**Comment by lihaoyi** [comment]
one request: can `scalacoverage` go under `contrib/`?

**Comment by lihaoyi** [comment]
yeah i have no idea of your dependency resolution issue; if you don't have time to figure it out we can close the PR and re-open it if someone wants to pick it up in future

**Comment by lihaoyi** [comment]
Closing due to inactivity

**Pull request closed without merging**